File: LeetCode/1358/main.cpp
class Solution {
public:
  int numberOfSubstrings(string s) {
    int ca = 0, cb = 0, cc = 0;
    int n = s.length(), left = 0, ans = 0;

    for (int i = 0; i < n; i++) {
      char ch = s[i];
      if (ch == 'a') {
        ca++;
      } else if (ch == 'b') {
        cb++;
      } else {
        cc++;
      }
      while (ca > 0 && cb > 0 && cc > 0) {
        ans += (n - i);
        if (s[left] == 'a') {
          ca--;
        } else if (s[left] == 'b') {
          cb--;
        } else {
          cc--;
        }
        left++;
      }
    }
    return ans;
  }
};

File: LeetCode/1358/main.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        ca, cb, cc = 0, 0, 0
        n, left, ans = len(s), 0, 0

        for i, ch in enumerate(s):
            if ch == "a":
                ca += 1
            elif ch == "b":
                cb += 1
            else:
                cc += 1

            while ca > 0 and cb > 0 and cc > 0:
                ans += n - i
                if s[left] == "a":
                    ca -= 1
                elif s[left] == "b":
                    cb -= 1
                else:
                    cc -= 1
                left += 1

        return ans
